DSMAJCEexamples/src/main/java/DSMAJCE.java

```diff
@@ -0,0 +1,221 @@
+            GCMParameterSpec gcmSpec = new GCMParameterSpec(tagLen, iv);
+            params.init(gcmSpec);
+            cipher.init(Cipher.DECRYPT_MODE, key, params);
+            cipher.updateAAD(AAD.getBytes());
+        } else
+            if (Objects.equals(mode, CTR)) {
+            cipher.init(Cipher.DECRYPT_MODE, key, new IvParameterSpec(iv));
+        } else{
+            params = cipher.getParameters();
+            cipher.init(Cipher.DECRYPT_MODE, key, params);
+        }
+        // decrypt the same content
+        return cipher.doFinal(cipherBytes);
+    }
+
+    private static byte[] tokenization(Provider provider, String mode, String padding, SecretKey key, String PLAIN) throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidAlgorithmParameterException, InvalidKeyException, IllegalBlockSizeException, BadPaddingException {
+        String algorithm = String.format("AES_%d/%s/%s", 256, mode, padding);
+        Cipher cipher = Cipher.getInstance(algorithm, provider);
+        AlgorithmParameters params = cipher.getParameters();
+        cipher.init(Cipher.ENCRYPT_MODE, key, params);
+        System.out.println("Plain text: " + PLAIN);
+        // Encrypted content
+        return cipher.doFinal(PLAIN.getBytes());
+    }
+    private static byte[] detokenization(Provider provider, String mode, String padding, SecretKey key, byte[] cipherBytes) throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidAlgorithmParameterException, InvalidKeyException, IllegalBlockSizeException, BadPaddingException {
+        String algorithm = String.format("AES_%d/%s/%s", 256, mode, padding);
+        Cipher cipher = Cipher.getInstance(algorithm, provider);
+        AlgorithmParameters params = cipher.getParameters();
+        // decrypt the same content
+        cipher.init(Cipher.DECRYPT_MODE, key, params);
+        return cipher.doFinal(cipherBytes);
+    }
+    public static byte[] sign(Provider provider, String algorithm, RSAPrivateKeyImpl rsaPrivateKey, String pssParameterDigest, MGF1ParameterSpec mgf1ParameterSpec) throws UnsupportedEncodingException, InvalidKeyException, SignatureException, NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        byte[] data = "test".getBytes("UTF8");
+        Signature sig = Signature.getInstance(algorithm, provider);
+        if (mgf1ParameterSpec != null || pssParameterDigest != null) {
+            sig.setParameter(new PSSParameterSpec(pssParameterDigest, "MGF1", mgf1ParameterSpec, 32, 1));
+        }
+        sig.initSign(rsaPrivateKey);
+        sig.update(data);
+        return sig.sign();
+    }
+    public static boolean verify(Provider provider, String algorithm, RSAPublicKeyImpl rsaPublicKey, byte[] signatureBytes, String pssParameterDigest, MGF1ParameterSpec mgf1ParameterSpec) throws UnsupportedEncodingException, InvalidKeyException, SignatureException, NoSuchAlgorithmException, InvalidAlgorithmParameterException {
```

Comment:
For sign and verify, we should pass in the data rather than create the signature data in the function.